### PR TITLE
MCS-704/705: Bring Addressables Catalog to TA1 Builds, and create a new custom registry file for them

### PR DESCRIPTION
**Pull request by JuanRMarquez-Ortiz**
A few things of interest here: 

- The variables on the Addressables Profiles have been streamlined so that switching build target is as simple as switching profile, instead of also requiring a switch of the profile variable in the `Addressable Group` settings. 
- There's now 3 Addressable Profiles. 
- `Remote` and `Development` remain, and their selection creates players that will load from the prod and dev buckets respectively. The build artifacts go to the same place they did before. The `Default` profile now does a local build, with the addressable files in the build folder itself. This was possible before, but it also required a profile change in the groups themselves, since each profile had not only local but also remote build/load paths. 
- - Note: On Linux the build has to be run with Root permissions, for some reason. Otherwise we get an error regarding connection to a server, which is odd as well. After bouncing it off of Dean, I decided to leave it as it is, so I can move to other stuff, and create a ticket to loop back to this to figure out why.

Second, when someone makes a player build without first building addressables, the build script will pick up the addressables catalogs from S3, so that when the TA1 team creates builds with their own custom objects they will still have access to whatever is on S3. 

I added a new custom object registry (to be created in `Resources/MCS/`), with a path reference in MCSMain, as the others. If it exists, it will be loaded along with the others.

When objects or assets are requested, we check Resources first, and only if the file isn't found there do we search in Addressables. As the addressables are searched with the prefix `Assets/Addressables/MCS/`, the ones in Resources will be looked for in `Resources/MCS/`. As so, the references in the registry file don't need this prefix, as it is added in code before the search.

There is room for improvement here, but it delves deeper into things that I am not sure I fully understand. If we add a new field to the object registry entries that define where they are, or otherwise separate the registries (I see that they are internally all combined into a single collection), we could then only search in the space that we expect them to be. For instance, I can look for the registry entry in the collection as per the name in the scene file, and then, based on their location, I can check only where they're supposed to be, instead of everywhere. This could give us a performance benefit. But I'm not sure which approach would be better. I don't know what the repercussions could be for each, or any others.

In any case, I believe this addresses the concerns Thomas brought up. if not, please let me know. ^_^

### CODE REVIEW

**Review by ThomasSchellenbergNextCentury — COMMENTED** [review]
Alright, so the TA1 teams can either add their new assets as addressables and rebuild the addressable asset bundles themselves (like Kyle and I were doing), OR add their new assets to the Resources folder, and either will work fine? That sounds super cool.

What's the best way for us to test this ourselves? Is it:
- Adding a new asset in the Addressables folder, marking it as addressable, changing the addressables profile to Default, rebuilding the addressable asset bundle locally, rebuilding the Unity app, and running the local build on a scene that should show the new asset
- Adding a new asset in the Resources folder, rebuilding the Unity app, and running the local build on a scene that show show the new asset

**Review by kdrumm — APPROVED** [review]
This looks good and works fine for me.  We should watch it the first time it goes through Jenkins to verify that works as we expect.  I still find it weird that sudo is needed for the 'default' build.  I'll probably use the localhost build and a local server as sudo and python didn't seem to like each other for me.  Could be my user error though.  Hopefully someone else can also test this soon.

**Comment by JuanRMarquez-Ortiz** [comment]
@ThomasSchellenbergNextCentury 
They can't build their own addressables. They will only do it as Resources. It's simpler for them, and it also assures that they will always pull our addressable assets, which include the private assets they can't directly access.

**Comment by ThomasSchellenbergNextCentury** [comment]
@JuanRMarquez-Ortiz 

> They can't build their own addressables.

That's completely fine, but could you explain why? I know that, if they (TA1) builds addressables locally, it won't include all of our (TA2's) private assets, but I thought Unity was supposed to fall back on the public asset bundle, if an asset's missing from the local bundle? Or is that not how it works?

**Comment by JuanRMarquez-Ortiz** [comment]
> > They can't build their own addressables.
> 
> That's completely fine, but could you explain why? I know that, if they (TA1) builds addressables locally, it won't include all of our (TA2's) private assets, but I thought Unity was supposed to fall back on the public asset bundle, if an asset's missing from the local bundle? Or is that not how it works?

I mean, they can build addressables, but if they do so when they next build the player it will use their addressables, not ours.

The original task as presented was to have multiple catalogs. After some discussions with @erick-u , we determined that we didn't need to use addressables with the TA1 team; they could instead use the resources folders that they already knew. That way their assets are bundled in their build, and when they built we could bring in the catalog data that we last built on the pipeline. When the system tries to spawn a GameObject, it will check our last built addressables. If the object isn't found, it will check the Resources folder where they place their assets. Only if both fail do we get an error that a given object wasn't found. 

We determined this process would be easier for them to us, and easier for us to implement. So we went with it.

**Comment by ThomasSchellenbergNextCentury** [comment]
@JuanRMarquez-Ortiz Thanks for the explanation. Your approach of using the Resources is fine in a general sense. However it isn't working for me in practice.

To make a new asset appear in one of our "scenes" (via the loaded JSON file data), you need to add an entry for it in one of the `*_object_registry.json` files. I'd think the process for this would be:
1. Copy an existing object registry file into the `Resources/MCS/` folder
2. Edit the copied file to add an entry for your new asset
3. Our code loads the modified file from the `Resources/MCS/` folder

However, since the object registry files already exist in our public addressable asset bundle, the code defaults to using the existing files, and not the modified file in the `Resources/MCS/` folder. (I can give you example steps to reproduce if you want them.)

I think we could solve this in a few ways:
1. Look for files in the Resources folder first, then fall back to using the public addressable asset bundle. I'm not sure what other ramifications this may cause though.
2. Ticket [MCS-704](https://nextcentury.atlassian.net/browse/MCS-704) involves creating a `custom_object_registry.json` file and reading from it in `MCSMain.Start()`. This file would presumably be located in the Resources folder. However, rather than having a blank/stub file committed in the repo, it would need to not exist (and the code would need to handle a non-existent file), or else we run into the same issue.

Maybe doing both would be best, unless there's a good reason to NOT do them both? I'd appreciate your thoughts. Thanks.

**Comment by JuanRMarquez-Ortiz** [comment]
@ThomasSchellenbergNextCentury 
Hmm. No. 2 sounds like the way to go, in regards to keeping as close to the existing procedures as possible. No. 1 is way easier to implement. 

i can do both, though. So I bring MCS-704 into this branch then and do it here as well? I don't mind. Shouldn't take long.

**Comment by ThomasSchellenbergNextCentury** [comment]
> So I bring MCS-704 into this branch then and do it here as well?

That's fine with me! :)

**Comment by JuanRMarquez-Ortiz** [comment]
OK, so I made some adjustments. I will modify the pull request description to include these. But for clarity of conversation, I'll explain here as well. 

Following @ThomasSchellenbergNextCentury 's comments above, I made the following changes:

I brought in MCS-704 into this pull request. I added a new custom object registry (to be created in `Resources/MCS/`), with a path reference in MCSMain, as the others. If it exists, it will be loaded along with the others. 

When objects or assets are loaded, we now check Resources first, and only if the file isn't found there do we search in Addressables. As the addressables are searched with the prefix `Assets/Addressables/MCS/`, the ones in Resources will be looked for in `Resources/MCS/`. As so, the references in the registry file don't need this prefix, as it is added in code before the search. 

There is room for improvement here, but it delves deeper into things that I am not sure I fully understand. If we add a new field to the object registry entries that define where they are, or otherwise separate the registries (I see that they are internally all combined into a single collection), we could then only search in the space that we expect them to be. For instance, I can look for the registry entry in the collection as per the name in the scene file, and then, based on their location, I can check only where they're supposed to be, instead of everywhere. This could give us a performance benefit. But I'm not sure which approach would be better. I don't know what the repercussions could be for each, or any others. 

In any case, I believe this addresses the concerns Thomas brought up. if not, please let me know. ^_^

**Comment by JuanRMarquez-Ortiz** [comment]
@ThomasSchellenbergNextCentury 
Also, your comment implies that they can override an object that exists in addressables, by creating a new object with the same name? Is this intended? Wouldn't that allow them to modify the scenes?

**Comment by ThomasSchellenbergNextCentury** [comment]
Great @JuanRMarquez-Ortiz . Thanks for giving us an update. I'll look at your changes presently.

> Also, your comment implies that they can override an object that exists in addressables, by creating a new object with the same name? Is this intended? Wouldn't that allow them to modify the scenes?

Yes, it would allow them to "modify" how the scenes are rendered, but only when using their local Unity build. However, I don't think we really care if they override our assets in their local builds, because our evaluation builds will stay the same. If they want to mess with something that already exists, then that's on them. (I guess the one caveat is if they mess something up accidentally, but hopefully they're smart enough to avoid that.) Pinging @deanwetherby for his input, but, as for me, I'm not particularly concerned. (Thanks for bringing it up, though! Still good to discuss.)

> I can look for the registry entry in the collection as per the name in the scene file, and then, based on their location, I can check only where they're supposed to be, instead of everywhere. This could give us a performance benefit. But I'm not sure which approach would be better. I don't know what the repercussions could be for each, or any others.

I understand what you're saying, and that might help a little, but I don't think it will help enough for us to worry about it at this time.

**Comment by kdrumm** [comment]
I would say we should be careful about letting them override on local builds.  There isn't anything preventing them from using local builds on an eval (I think).  Clearly they shouldn't and maybe we should verify.  For the record, opics used the auto downloaded version which is what I'd hope everyone would use for eval.

**Comment by ThomasSchellenbergNextCentury** [comment]
> There isn't anything preventing them from using local builds on an eval

Yeah, if this is the case, then it seems like a bug we should handle on our end.

**Comment by kdrumm** [comment]
> > There isn't anything preventing them from using local builds on an eval
> 
> Yeah, if this is the case, then it seems like a bug we should handle on our end.

Its not a bug per say.  They just setup their software and the AWS instance how they want.  There might be a way to check or prevent it automatically.  I'm just not sure how at the moment.

**Comment by JuanRMarquez-Ortiz** [comment]
@ThomasSchellenbergNextCentury  @kdrumm 
Not just modify a prefab, but through it's code, modify the entire scene, which is awfully powerful. 

there's a few ways we can restrict this, such as checking addressables before resources, or checking if a given name exists on both, then use addressables? We can definitely come up with something if it's important. 

As for the perf consideration, object load is something that is done AT LEAST tens of times in each scene. And you guys run, what, thousands? Even a second less on each adds up to 17 minutes.... 

Yeah, nevermind. Maybe not that huge a deal. Still, feels sloppy, I guess. At the very least, it's something that may be worth noting down, in case we do notice perf variations running the code before and after. I couldn't notice much, but I'm not running the volume you guys are.

**Comment by ThomasSchellenbergNextCentury** [comment]
@JuanRMarquez-Ortiz 

> Not just modify a prefab, but through it's code, modify the entire scene, which is awfully powerful.

Do you mean, like, by adding a custom script to it? They could do that with a brand new prefab too, right? It's been our mentality for a while now that we're fine with giving them the ability to break things themselves (with the proper guidance and documentation on how to avoid doing so), and I feel like this is just an extension of that mentality, so I still don't have any problem with it. (I'm happy to debate it though.)

> As for the perf consideration, object load is something that is done AT LEAST tens of times in each scene.

Why 10+ times? Is this just something we're not handling well in our own code? Maybe this would be a better area for performance improvements?

@kdrumm 

> Its not a bug per say. They just setup their software and the AWS instance how they want. There might be a way to check or prevent it automatically. I'm just not sure how at the moment.

Yeah, I really think this is a truly bad idea. We should have an environment variable or something to override their setting. What if, like the NYU team, they were trying to use an old version like 0.4.3 that just wouldn't work at all? Seems like something we should fix in Eval 5, IMO.

**Review by ThomasSchellenbergNextCentury — APPROVED** [review]
Looks good overall. A few comments:
- Using the `custom_object_registry` in the `Resources` folder works for me. However, you can't override any of the other object registry files (with versions in the `Resources` folder) because we always read them directly from the addressables. I think this is fine -- I can't think of any realistic use cases in which this would be needed -- so no action is necessary.
- If you try to override an existing object "type" (like "soccer_ball") via the `custom_object_registry` file, the Dictionary.Add function will error because the key already exists. I also think this is fine (though it would probably also be fine the other way) -- just noting this behavior for our team's awareness.
- Overriding other existing assets (like adding `Resources/MCS/Soccer_Ball/Prefabs/Soccer Ball.prefab` to switch the prefab used whenever making a "soccer ball") works, which is fine. (I won't rehash the conversation we already had about this.)
- For sake of clarity, I think it's better to have an empty "objects" array in the `custom_object_registry` file, which I've added and pushed myself. (I'm happy to discuss it if there's disagreement.) We'll eventually need to decide what to add to the README, and how that compares to what's already in the [existing documentation](https://nextcenturycorporation.github.io/MCS/models.html), but we can worry about all that later.

In summary, approved. Thanks for all your hard work, Juan! I know the back-and-forth is cumbersome.

**Comment by JuanRMarquez-Ortiz** [comment]
@ThomasSchellenbergNextCentury 

- I will make the Disctionary.add error more explicit and clear on the log, removing the error by checking first and reporting on the log if there's a duplicate. Feels safer and cleaner, and is more instructive. 

Everything else sounds good. Let me know when/if to merge.
